File: Chessko/ViewModels/LearnViewModel.swift
import SwiftUI

// MARK: - Learn Scenario

enum LearnScenario: String, Equatable {
    case castling
    case enPassant
    case promotion

    var label: String {
        switch self {
        case .castling:  return Loc("Rokada")
        case .enPassant: return "En passant"
        case .promotion: return Loc("Promocija")
        }
    }
}

// MARK: - Learn View Model
//
// Manages an interactive "sandbox" board for the tutorial tab.
// The player can freely tap any square to relocate the selected piece
// and see its reachable squares highlighted.

@Observable
@MainActor
final class LearnViewModel {

    // MARK: - State

    var selectedPieceType: PieceType = .queen
    var piecePosition: Position = Position(row: 4, col: 3)   // d4 — good center default
    var activeScenario: LearnScenario? = nil

    // Derived board + moves (refreshed on every state change)
    private(set) var board: [[ChessPiece?]] = emptyBoard()
    private(set) var legalMoves: [ChessMove] = []

    // BoardView compatibility (animation stubs — tutorial doesn't animate)
    var animatingPiece: AnimatingPiece? = nil
    var flyingCapture:  FlyingCapture?  = nil
    var lastMove:       ChessMove?      = nil

    // BoardView reads this to draw the gray "selected" highlight
    var selectedPosition: Position? { piecePosition }
    var isPlayerTurn: Bool { true }

    // MARK: - Init

    init() { refresh() }

    // MARK: - Piece Selection

    func select(piece: PieceType) {
        selectedPieceType = piece
        activeScenario    = nil
        piecePosition     = defaultPosition(for: piece)
        refresh()
    }

    // MARK: - Scenario

    /// Available extra-scenarios for the currently selected piece.
    var availableScenarios: [LearnScenario] {
        switch selectedPieceType {
        case .king:  return [.castling]
        case .pawn:  return [.enPassant, .promotion]
        default:     return []
        }
    }

    func toggleScenario(_ scenario: LearnScenario) {
        if activeScenario == scenario {
            // Deactivate → go back to normal piece view
            activeScenario = nil
            piecePosition  = defaultPosition(for: selectedPieceType)
        } else {
            activeScenario = scenario
            piecePosition  = scenarioStartPosition(scenario)
        }
        refresh()
    }

    // MARK: - Tap Handler

    /// User tapped a square on the board.
    /// • Tapping the piece's own square → no-op
    /// • Tapping anything else → relocate piece (free placement)
    func tap(position: Position) {
        guard position != piecePosition else { return }
        // Exit special scenario when user freely moves the piece
        if activeScenario != nil { activeScenario = nil }
        piecePosition = position
        refresh()
    }

    // MARK: - Info text

    var infoTitle: String {
        switch activeScenario {
        case .none:        return selectedPieceType.srbName
        case .castling:    return Loc("Rokada")
        case .enPassant:   return "En passant"
        case .promotion:   return Loc("Promocija")
        }
    }

    var infoText: String {
        switch activeScenario {
        case .none:
            switch selectedPieceType {
            case .pawn:
                return Loc("Kreće se jedno polje napred. Uzima figuru dijagonalno ispred sebe. Sa startne pozicije može da skoči i dva polja odjednom.")
            case .rook:
                return Loc("Kreće se horizontalno ili vertikalno, koliko god polja želi. Ne može preskakati figure.")
            case .knight:
                return Loc("Kreće se u obliku slova L: dva polja u jednom pravcu pa jedno bočno. Jedina figura koja može preskočiti druge.")
            case .bishop:
                return Loc("Kreće se dijagonalno, koliko god polja želi. Uvek ostaje na istoj boji polja.")
            case .queen:
                return Loc("Najjača figura na tabli. Kombinuje kretanje topa i lovca — horizontalno, vertikalno i dijagonalno.")
            case .king:
                return Loc("Kreće se jedno polje u bilo kom smeru. Ne sme stati na polje koje napada protivnik. Zaštiti ga!")
            }
        case .castling:
            return Loc("Poseban potez: ako kralj i top nisu se još pomerali i između njih nema figura, kralj skoči dva polja ka topu, a top preskoči kralja. Tapni g1 (kratka rokada) ili c1 (duga rokada).")
        case .enPassant:
            return Loc("Posebno uzimanje pešakom: ako protivnički pešak skoči dva polja i nađe se pored tvojeg pešaka, možeš ga uzeti 'u prolazu' — kao da se pomerio samo jedno polje. Tapni d6.")
        case .promotion:
            return Loc("Kad beli pešak stigne do osmog reda (redovi 8), može se pretvoriti u bilo koju figuru — gotovo uvek u damu. Tapni e8.")
        }
    }

    var movesCountLabel: String {
        let n = legalMoves.count
        switch n {
        case 0:  return Loc("Nema mogućih poteza")
        case 1:  return Loc("1 mogući potez")
        default: return LocF("%lld mogućih poteza", n)
        }
    }

    // MARK: - Private

    private func refresh() {
        let (b, state) = buildBoardAndState()
        board      = b
        legalMoves = MoveGenerator.legalMoves(for: .white, in: state)
            .filter { $0.from == piecePosition }
    }

    private func buildBoardAndState() -> ([[ChessPiece?]], GameState) {
        var b = Self.emptyBoard()
        var castleKS = false
        var castleQS = false
        var epTarget: Position? = nil

        switch activeScenario {

        case .none:
            // Place the tutorial piece
            b[piecePosition.row][piecePosition.col] = ChessPiece(type: selectedPieceType, color: .white)
            // For pawns, add dummy enemy pieces on the capture diagonals so captures are visible
            if selectedPieceType == .pawn {
                let capRow = piecePosition.row - 1   // white pawn moves up (row decreases)
                if capRow >= 0 {
                    if piecePosition.col > 0 {
                        b[capRow][piecePosition.col - 1] = ChessPiece(type: .pawn, color: .black)
                    }
                    if piecePosition.col < 7 {
                        b[capRow][piecePosition.col + 1] = ChessPiece(type: .pawn, color: .black)
                    }
                }
            }

        case .castling:
            // King on e1 + both rooks — both castling rights active
            b[7][4] = ChessPiece(type: .king, color: .white)   // e1
            b[7][0] = ChessPiece(type: .rook, color: .white)   // a1
            b[7][7] = ChessPiece(type: .rook, color: .white)   // h1
            castleKS = true
            castleQS = true

        case .enPassant:
            // White pawn on e5; black pawn on d5 that "just moved" two squares
            b[3][4] = ChessPiece(type: .pawn, color: .white)   // e5
            b[3][3] = ChessPiece(type: .pawn, color: .black)   // d5
            epTarget = Position(row: 2, col: 3)                 // d6 — en passant target square

        case .promotion:
            // White pawn one step from queening
            b[1][4] = ChessPiece(type: .pawn, color: .white)   // e7
        }

        let state = GameState(
            board: b,
            currentTurn: .white,
            whiteCanCastleKingside:  castleKS,
            whiteCanCastleQueenside: castleQS,
            blackCanCastleKingside:  false,
            blackCanCastleQueenside: false,
            enPassantTarget: epTarget,
            moveHistory: [],
            capturedByWhite: [],
            capturedByBlack: [],
            status: .playing
        )
        return (b, state)
    }

    // MARK: - Helpers

    private func defaultPosition(for piece: PieceType) -> Position {
        switch piece {
        case .pawn:  return Position(row: 6, col: 4)   // e2 — shows double-move + captures
        case .king:  return Position(row: 4, col: 4)   // e4
        default:     return Position(row: 4, col: 3)   // d4 — nice centre for max mobility
        }
    }

    private func scenarioStartPosition(_ scenario: LearnScenario) -> Position {
        switch scenario {
        case .castling:  return Position(row: 7, col: 4)   // e1
        case .enPassant: return Position(row: 3, col: 4)   // e5
        case .promotion: return Position(row: 1, col: 4)   // e7
        }
    }

    private static func emptyBoard() -> [[ChessPiece?]] {
        Array(repeating: Array(repeating: nil, count: 8), count: 8)
    }
}

// MARK: - Preview

#Preview {
    NavigationStack {
        LessonDetailView(lesson: LessonInfo.all[0], pieceExplorer: LearnViewModel())
    }
}

File: ChesskoAndroid/app/src/main/java/com/veljkoni/chessko/viewmodels/LearnViewModel.kt
package com.veljkoni.chessko.viewmodels

import com.veljkoni.chessko.logic.loc
import com.veljkoni.chessko.logic.locF

import androidx.compose.runtime.*
import com.veljkoni.chessko.logic.MoveGenerator
import com.veljkoni.chessko.models.*

enum class LearnScenario(val label: String) {
    CASTLING(loc("Rokada")),
    EN_PASSANT("En passant"),
    PROMOTION(loc("Promocija"))
}

class LearnViewModel {
    var selectedPieceType by mutableStateOf(PieceType.QUEEN)
    var piecePosition by mutableStateOf(Position(4, 3)) // d4 default
    var activeScenario by mutableStateOf<LearnScenario?>(null)

    var board by mutableStateOf<List<List<ChessPiece?>>>(emptyBoard())
        private set
    var legalMoves by mutableStateOf<List<ChessMove>>(emptyList())
        private set

    val selectedPosition: Position?
        get() = piecePosition

    val isPlayerTurn: Boolean
        get() = true

    init {
        refresh()
    }

    fun select(piece: PieceType) {
        selectedPieceType = piece
        activeScenario = null
        piecePosition = defaultPosition(piece)
        refresh()
    }

    val availableScenarios: List<LearnScenario>
        get() = when (selectedPieceType) {
            PieceType.KING -> listOf(LearnScenario.CASTLING)
            PieceType.PAWN -> listOf(LearnScenario.EN_PASSANT, LearnScenario.PROMOTION)
            else -> emptyList()
        }

    fun toggleScenario(scenario: LearnScenario) {
        if (activeScenario == scenario) {
            activeScenario = null
            piecePosition = defaultPosition(selectedPieceType)
        } else {
            activeScenario = scenario
            piecePosition = scenarioStartPosition(scenario)
        }
        refresh()
    }

    fun tap(position: Position) {
        if (position == piecePosition) return
        if (activeScenario != null) {
            activeScenario = null
        }
        piecePosition = position
        refresh()
    }

    val infoTitle: String
        get() = when (activeScenario) {
            null -> selectedPieceType.srbName
            LearnScenario.CASTLING -> loc("Rokada")
            LearnScenario.EN_PASSANT -> "En passant"
            LearnScenario.PROMOTION -> loc("Promocija")
        }

    val infoText: String
        get() = when (activeScenario) {
            null -> when (selectedPieceType) {
                PieceType.PAWN -> loc("Kreće se jedno polje napred. Uzima figuru dijagonalno ispred sebe. Sa startne pozicije može da skoči i dva polja odjednom.")
                PieceType.ROOK -> loc("Kreće se horizontalno ili vertikalno, koliko god polja želi. Ne može preskakati figure.")
                PieceType.KNIGHT -> loc("Kreće se u obliku slova L: dva polja u jednom pravcu pa jedno bočno. Jedina figura koja može preskočiti druge.")
                PieceType.BISHOP -> loc("Kreće se dijagonalno, koliko god polja želi. Uvek ostaje na istoj boji polja.")
                PieceType.QUEEN -> loc("Najjača figura na tabli. Kombinuje kretanje topa i lovca — horizontalno, vertikalno i dijagonalno.")
                PieceType.KING -> loc("Kreće se jedno polje u bilo kom smeru. Ne sme stati na polje koje napada protivnik. Zaštiti ga!")
            }
            LearnScenario.CASTLING -> loc("Poseban potez: ako kralj i top nisu se još pomerali i između njih nema figura, kralj skoči dva polja ka topu, a top preskoči kralja. Tapni g1 (kratka rokada) ili c1 (duga rokada).")
            LearnScenario.EN_PASSANT -> loc("Posebno uzimanje pešakom: ako protivnički pešak skoči dva polja i nađe se pored tvojeg pešaka, možeš ga uzeti 'u prolazu' — kao da se pomerio samo jedno polje. Tapni d6.")
            LearnScenario.PROMOTION -> "Kad beli pešak stigne do osmog reda, može se pretvoriti u bilo koju figuru — gotovo uvek u damu. Tapni e8."
        }

    val movesCountLabel: String
        get() {
            val n = legalMoves.size
            return when (n) {
                0 -> loc("Nema mogućih poteza")
                1 -> loc("1 mogući potez")
                else -> locF("%d mogućih poteza", n)
            }
        }

    private fun refresh() {
        val (b, state) = buildBoardAndState()
        board = b
        legalMoves = MoveGenerator.legalMoves(PieceColor.WHITE, state)
            .filter { it.from == piecePosition }
    }

    private fun buildBoardAndState(): Pair<List<List<ChessPiece?>>, GameState> {
        val b = MutableList(8) { MutableList<ChessPiece?>(8) { null } }
        var castleKS = false
        var castleQS = false
        var epTarget: Position? = null

        when (activeScenario) {
            null -> {
                b[piecePosition.row][piecePosition.col] = ChessPiece(selectedPieceType, PieceColor.WHITE)
                if (selectedPieceType == PieceType.PAWN) {
                    val capRow = piecePosition.row - 1
                    if (capRow >= 0) {
                        if (piecePosition.col > 0) {
                            b[capRow][piecePosition.col - 1] = ChessPiece(PieceType.PAWN, PieceColor.BLACK)
                        }
                        if (piecePosition.col < 7) {
                            b[capRow][piecePosition.col + 1] = ChessPiece(PieceType.PAWN, PieceColor.BLACK)
                        }
                    }
                }
            }
            LearnScenario.CASTLING -> {
                b[7][4] = ChessPiece(PieceType.KING, PieceColor.WHITE)
                b[7][0] = ChessPiece(PieceType.ROOK, PieceColor.WHITE)
                b[7][7] = ChessPiece(PieceType.ROOK, PieceColor.WHITE)
                castleKS = true
                castleQS = true
            }
            LearnScenario.EN_PASSANT -> {
                b[3][4] = ChessPiece(PieceType.PAWN, PieceColor.WHITE)
                b[3][3] = ChessPiece(PieceType.PAWN, PieceColor.BLACK)
                epTarget = Position(2, 3)
            }
            LearnScenario.PROMOTION -> {
                b[1][4] = ChessPiece(PieceType.PAWN, PieceColor.WHITE)
            }
        }

        val state = GameState(
            board = b.map { it.toList() }.toList(),
            currentTurn = PieceColor.WHITE,
            whiteCanCastleKingside = castleKS,
            whiteCanCastleQueenside = castleQS,
            blackCanCastleKingside = false,
            blackCanCastleQueenside = false,
            enPassantTarget = epTarget,
            moveHistory = emptyList(),
            capturedByWhite = emptyList(),
            capturedByBlack = emptyList(),
            status = GameStatus.Playing
        )
        return Pair(state.board, state)
    }

    private fun defaultPosition(piece: PieceType): Position {
        return when (piece) {
            PieceType.PAWN -> Position(6, 4) // e2
            PieceType.KING -> Position(4, 4) // e4
            else -> Position(4, 3) // d4
        }
    }

    private fun scenarioStartPosition(scenario: LearnScenario): Position {
        return when (scenario) {
            LearnScenario.CASTLING -> Position(7, 4) // e1
            LearnScenario.EN_PASSANT -> Position(3, 4) // e5
            LearnScenario.PROMOTION -> Position(1, 4) // e7
        }
    }

    companion object {
        fun emptyBoard(): List<List<ChessPiece?>> {
            return List(8) { List<ChessPiece?>(8) { null } }
        }
    }
}
